File: p437.cpp
#include <bits/stdc++.h>
#include <TreeNode.hpp>

using namespace std;

class Solution {
public:
    int find(TreeNode* root, int &sum, int accumulate) {
        if(root == NULL) {
            return 0;
        }
        accumulate += root->val;
        if(accumulate == sum) {
            return 1 + find(root->left, sum, accumulate) + find(root->right, sum, accumulate);
        }
        else {
            return find(root->left, sum, accumulate) + find(root->right, sum, accumulate);
        }
    }
    int pathSum(TreeNode* root, int sum) {
        if(root == NULL) {
            return 0;
        }
        return find(root, sum, 0) + pathSum(root->left, sum) + pathSum(root->right, sum);
    }
};

File: p437.py
from DS.TreeNode import TreeNode

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        if root is None:
            return 0

        def dfs(node, value):
            if node is None:
                return 0
            value += node.val
            if value == sum:
                return 1 + dfs(node.left, value) + dfs(node.right, value)
            return dfs(node.left, value) + dfs(node.right, value)

        return dfs(root, 0) + self.pathSum(root.left, sum) + self.pathSum(root.right, sum)
